StackExRedis/NHibernate.Caches.StackExRedis/RedisCacheProvider.cs

```diff
@@ -0,0 +1,232 @@
+
+		private ConnectionMultiplexer _connectionMultiplexer;
+		private RedisCacheConfiguration _defaultCacheConfiguration;
+
+		static RedisCacheProvider()
+		{
+			Log = NHibernateLogger.For(typeof(RedisCacheProvider));
+			ConfiguredCacheRegions = new Dictionary<string, RegionConfig>();
+
+			if (!(ConfigurationManager.GetSection("redis") is CacheConfig config))
+				return;
+
+			ConfiguredCache = config;
+			foreach (var cache in config.Regions)
+			{
+				ConfiguredCacheRegions.Add(cache.Region, cache);
+			}
+		}
+
+		/// <inheritdoc />
+		public ICache BuildCache(string regionName, IDictionary<string, string> properties)
+		{
+			if (regionName == null)
+			{
+				regionName = string.Empty;
+			}
+
+			var regionConfiguration = ConfiguredCacheRegions.TryGetValue(regionName, out var regionConfig) 
+				? BuildRegionConfiguration(regionConfig, properties) 
+				: BuildRegionConfiguration(regionName, properties);
+
+			Log.Debug("Building cache: {0}", regionConfiguration.ToString());
+
+			return BuildCache(_defaultCacheConfiguration, regionConfiguration, properties);
+		}
+
+		/// <inheritdoc />
+		public long NextTimestamp()
+		{
+			return Timestamper.Next();
+		}
+
+		/// <inheritdoc />
+		public void Start(IDictionary<string, string> properties)
+		{
+			var configurationString = GetString(RedisEnvironment.Configuration, properties, ConfiguredCache?.Configuration);
+			if (string.IsNullOrEmpty(configurationString))
+			{
+				throw new CacheException("The StackExchange.Redis configuration string was not provided.");
+			}
+
+			Log.Debug("Starting with configuration string: {0}", configurationString);
+
+			_defaultCacheConfiguration = BuildDefaultConfiguration(properties);
+
+			Log.Debug("Default configuration: {0}", _defaultCacheConfiguration);
+
+			TextWriter textWriter = Log.IsDebugEnabled() ? new NHibernateTextWriter(Log) : null;
+			Start(configurationString, properties, textWriter);
+		}
+
+		/// <inheritdoc />
+		public virtual void Stop()
+		{
+			try
+			{
+				if (Log.IsDebugEnabled())
+				{
+					Log.Debug("Releasing connection.");
+				}
+				_connectionMultiplexer.Dispose();
+				_connectionMultiplexer = null;
+			}
+			catch (Exception e)
+			{
+				Log.Error(e, "An error occurred while releasing the connection.");
+			}
+		}
+
+		/// <summary>
+		/// Callback to perform any necessary initialization of the underlying cache implementation
+		/// during ISessionFactory construction.
+		/// </summary>
+		/// <param name="configurationString">The Redis configuration string.</param>
+		/// <param name="properties">NHibernate configuration settings.</param>
+		/// <param name="textWriter">The text writer.</param>
+		protected virtual void Start(string configurationString, IDictionary<string, string> properties, TextWriter textWriter)
+		{
+			var configuration = ConfigurationOptions.Parse(configurationString);
+			_connectionMultiplexer = ConnectionMultiplexer.Connect(configuration, textWriter);
```

Comment:
I think it's great that the ConnectionMultiplexer can be configured and instantiated like this. However, ConnectionMultiplexers should be "as singleton as possible". It would be nice to have a way to provide it programmatically, if there is already an instance that you want to use. This could be done like in the other provider (NHibernate.Caches.Redis), which has a static SetConnectionMultiplexer method, but even better would be a callback. Another callback could be used intercept GetDatabase.

Reply:
In my opinion the right way to go is to have an additional constructor that takes the `ConnectionMultiplexer` as a parameter. I know that NHibernate currently does not have a [simple way to set the IObjectsFactory](https://github.com/nhibernate/nhibernate-core/issues/1671) which would allow us to use our favourite IoC container. My plan would be to provide a PR to decuple the `IObjectsFactory` from `BytecodeProvider` and then add the extra constructor with the `ConnectionMultiplexer` if you agree? This is an example how the final solution would look like:
```
container.RegisterSingleton(myConnectionMultiplexer);
Cfg.Environment.ObjectsFactory = new MyObjectsFactory(container);
... // build a session factory
// The constructor with the ConnectionMultiplexer parameter would be called by the IoC container
```
For `GetDatabase` I would rather provide a `IRedisDatabaseProvider` interface that would look like:
```
public interface IRedisDatabaseProvider {
  IDatabase GetDatabase(ConnectionMultiplexer connectionMultiplexer, int database)
}
```
which could be configured in the configuration.

Reply:
Yes, if the IoC experience gets better (there's so much static stuff in the initialization...), that would be a good way. Probably using IConnectionMultiplexer, instead of ConnectionMultiplexer, though.